StitchCore/StitchCore/Source/Core/StitchClient.swift

```diff
@@ -268,27 +276,41 @@ public class StitchClient: StitchClientType {
     public func login(withProvider provider: AuthProvider) -> Promise<UserId> {
         self.authProvider = provider
 
-        if isAuthenticated, let auth = auth {
-            printLog(.info, text: "Already logged in, using cached token.")
-            return Promise.init(value: auth.userId)
+        func doLoginRequest() -> Promise<UserId> {
+            return httpClient.doRequest {
+                $0.method = .post
+                $0.endpoint = self.routes.authProvidersLoginRoute(provider: provider.type)
+                $0.isAuthenticatedRequest = false
+                try $0.encode(withData: self.getAuthRequest(provider: provider))
+                }.flatMap { [weak self] any in
+                    guard let strongSelf = self else { throw StitchError.clientReleased }
+                    let authInfo = try JSONDecoder().decode(AuthInfo.self,
+                                                            from: JSONSerialization.data(withJSONObject: any))
+                    strongSelf.httpClient.authInfo = authInfo
+                    strongSelf._auth = Auth(stitchClient: strongSelf,
+                                            stitchHttpClient: strongSelf.httpClient,
+                                            userId: authInfo.userId)
+                    strongSelf.onLogin()
+                    return authInfo.userId
+                }
         }
 
-        return httpClient.doRequest {
-            $0.method = .post
-            $0.endpoint = self.routes.authProvidersLoginRoute(provider: provider.type)
-            $0.isAuthenticatedRequest = false
-            try $0.encode(withData: self.getAuthRequest(provider: provider))
-        }.flatMap { [weak self] any in
-            guard let strongSelf = self else { throw StitchError.clientReleased }
-            let authInfo = try JSONDecoder().decode(AuthInfo.self,
-                                                    from: JSONSerialization.data(withJSONObject: any))
-            strongSelf.httpClient.authInfo = authInfo
-            strongSelf._auth = Auth(stitchClient: strongSelf,
-                                    stitchHttpClient: strongSelf.httpClient,
-                                    userId: authInfo.userId)
-            strongSelf.onLogin()
-            return authInfo.userId
+        if isAuthenticated, let auth = auth {
```

Comment:
[nit] Easier to return early in a small block of code checking `if !isAuthenticated`.

[style?] In fact, it must be the case that auth is not `nil`, which I think means we can use the force unwrap expression `!auth.userId`. @jsflax, as the owners of the state of `auth`, is it fair to say we want a runtime error to happen if `auth` is `nil` when `isAuthenticated` is true or is it too dangerous.

Reply:
Holding off on @jsflax 's review before making changes here. But if we agree that a runtime error is okay then I can change this to do a `guard isAuthenticated else { return doLoginRequest() }` block instead of the big `if` block.

Reply:
Conditional bindings also allow inner properties of structures, so we could just strip out `userId`. Adam's idea of using a `guard` here also works.